Dashboard: Add unit test

proy30 opened the pull request:
This first CI test for the dashboard utilizes inputs from `examples/run_fodo.py`. The `seleniumbase` python library is utilized to simulate user interaction with the dashboard.

Passes the following workflows:
- [X] macOS
- [X] Ubuntu
- [X] Stubs

I currently have the dashboard tests run **separately** from the other testing. The initial reason was to isolate the job in the workflows for trial and error.

One note for the code changes is that `run installed python module` in `macOS` and `Ubuntu` is moved to run before both the dashboard and unit testing, rather than after (which was the previous state). This is because the dashboard tests rely on the python bindings for specific imports.

Closes #706 
Closes #667 

Because the `Windows` and `Tooling` workflows do not pass, and possibly is a more involved process to solve, the following issue is created https://github.com/BLAST-ImpactX/impactx/issues/968

proy30 committed:
add: unit test

proy30 committed:
test_commit: separate dashboard tests from main test suite

applying to ubunto.yml, macos.yml and windows.yml to test

proy30 committed:
apply pre-commit changes

proy30 committed:
update: helper functions for code clarity

proy30 committed:
fix: windows and tooling yml

proy30 committed:
add back removed try/except

important to keep so that it retries the js script

proy30 committed:
remove @staticmethod tag in unneeded function

caught by ci

proy30 committed:
update: set_state

proy30 committed:
set inputs for csr/isr to False

proy30 committed:
remove tooling and windows

because they do not work

proy30 committed:
remove dashboard test in msvc workflow

proy30 committed:
correctly separate dashboard tests

proy30 committed:
separate dashboard test

proy30 committed:
simplify test_dashboard

proy30 committed:
add set_input

to combine usage of set_state and set_js, so that it is less confusing to use

proy30 committed:
simplify test

proy30 committed:
Merge branch 'development' into dashboard/add_first_unit_test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
fix precomit

proy30 committed:
Merge branch 'dashboard/add_first_unit_test' of github.com:proy30/impactx into dashboard/add_first_unit_test

proy30 committed:
fix type hint for stubs

proy30 committed:
clean up docstring in test_dashboard

not needed to mention anymore as we use set_input instead

proy30 committed:
Merge branch 'development' into dashboard/add_first_unit_test

proy30 committed:
fix stub

EZoni commented on .github/workflows/ubuntu.yml:
Here I see that the dashboard tests are added as a step in each of the jobs of this workflow. So we run the dashboard tests for the job "GCC w/ MPI w/ Python" and also for the job "GCC w/o MPI w/ Python".

What differences should one expect from interacting with the dashboard in these two different configurations? 

Somehow I would have expected the dashboard tests to be within a new, separate job in this workflow, rather than a new step in each existing job. May be something to discuss in person.

proy30 replied:
Currently, the dashboard does not operate differently in any build, including Ubuntu. The dashboard tests were added to the same jobs that ran our unit tests. As we discussed, it would make sense that future dashboard versions would be different (testing specific inputs for MPI on `GCC w/ MPI w/ Python` only) and could utilize the current implementation of CI. 

Would need input from @ax3l.

ax3l replied:
Resolved in last meeting :+1: 

Summary: tests are integrated with pytest (great) and do not run long at the moment (<1min). We exclude them nonetheless already for a few runners where we also exclude "slow" unit tests, because we expect them to grow.

EZoni commented:
Is there a place in the documentation where we could add guidance on how to run the dashboard tests manually? Including as simple as `ctest --test-dir build --output-on-failure -R dashboard` or similar.

EZoni commented:
Are all the changes in src/python, as opposed to the changes in tests/python, necessary for the tests to work? In other words, do all those changes belong to this PR or could any of those be implemented in a separate PR, just for clarity?

proy30 commented:
> Are all the changes in src/python, as opposed to the changes in tests/python, necessary for the tests to work? In other words, do all those changes belong to this PR or could any of those be implemented in a separate PR, just for clarity?

@EZoni Yes, the addition of the `id` is needed by selenium to interact with the specific elements. Everything else (the removal of the `@staticmethod` and fixing of the type hint) was caught by the Stubs CI.

> Is there a place in the documentation where we could add guidance on how to run the dashboard tests manually? Including as simple as ctest --test-dir build --output-on-failure -R dashboard or similar.

@EZoni  Yes, on a TODO to add to https://github.com/BLAST-ImpactX/impactx/issues/983

EZoni commented on tests/python/dashboard/test_dashboard.py:
```diff
@@ -0,0 +1,124 @@
+    importlib.util.find_spec("seleniumbase") is None,
+    reason="seleniumbase is not available",
+)
+def test_dashbnoard():
```
```suggestion
def test_dashboard():

```

EZoni commented on tests/python/dashboard/utils.py:
```diff
@@ -0,0 +1,220 @@
+            return
+
+
+def wait_for_server_ready(process, timeout=TIMEOUT):
+    """
+    Waits for the server of the dashboard to launch.
+    """
+    for i in range(timeout):
+        print(f"Waiting for dashboard server to start - {i}s elapsed.")
+        line = process.stdout.readline()
+        if line:
+            print(line, end="")
+            if "App running at:" in line:
+                for _ in range(2):
+                    next_line = process.stdout.readline()
+                    if next_line:
+                        print(next_line, end="")
+                print("Dashboard server has successfully launched.")
+                return
+        else:
+            time.sleep(1)
+
+    raise RuntimeError(f"Dashboard server did not start within {timeout} seconds.")
```
Here I wonder if one could make use of Trame's built-in server control, e.g. `ctrl.on_server_ready` or something like https://trame.readthedocs.io/en/latest/core.server.html#trame_server.core.Server.ready or https://trame.readthedocs.io/en/latest/core.server.html#trame_server.core.Server.get_server_state.

Or, somewhat related, as you check `.trame__loader` above, on server ready `dir(state)` gives
```
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_change_callbacks', '_children_state', '_hot_reload', '_modified_keys', '_parent_state', '_pending_update', '_push_state_fn', '_pushed_state', '_ready_flag', '_state_listeners', '_translator', 'change', 'clean', 'client_only', 'dirty', 'flush', 'has', 'initial', 'is_dirty', 'is_dirty_all', 'is_ready', 'modified_keys', 'ready', 'setdefault', 'to_dict', 'translator', 'update']
```
I see `is_ready` and `ready` - I wonder if one can access those through `sb` somehow.

Some version of the above may eventually lead to a simpler implementation than parsing the `stdout`.

ax3l replied:
ping @proy30 did you see this comment?

proy30 replied:
@EZoni, Definitely think there is a simpler alternative like your suggestion. I have not experimented with Trames server control, so I would not know for sure now. A new TODO on my end to see how it works.

EZoni reviewed (commented):
A couple more comments!

proy30 committed:
fix test name

Co-authored-by: Edoardo Zoni <59625522+EZoni@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 commented on .github/workflows/ubuntu.yml:
```diff
@@ -117,9 +122,20 @@ jobs:
           -DImpactX_PYTHON=ON
         cmake --build build -j 4
 
+    - name: run installed python module
+      run: |
+        cmake --build build --target pip_install
+
+        python3 examples/fodo/run_fodo.py
+
+    - name: run dashboard tests
+      run: |
+        cd tests/python/dashboard
+        pytest --confcutdir=.
+
     - name: run tests
```
@ax3l Moved `run installed python module` to run before testing.

ax3l replied:
I improved the CMake now, so we do not need this changed -- restored to previous state :)

ax3l commented on .github/workflows/ubuntu.yml:
```diff
@@ -53,9 +53,20 @@ jobs:
 
+    - name: run installed python module
+      run: |
+        cmake --build build --target pip_install
```
Double check:
- As of now, do pytest test work even without `pip_install` (yes?)?
- Can we make the dashboard tests work the same way?

ax3l replied:
(To be revisited in a follow-up.)

ax3l replied:
I will add this to this PR now.

proy30 committed:
remove dashboard-specific job from macos, ubuntu w/MPI, stubs

proy30 committed:
Merge branch 'development' into dashboard/add_first_unit_test

ax3l commented:
@proy30 sorry, small conflict in `tests/python/requirements.txt`

proy30 committed:
Merge branch 'development' into dashboard/add_first_unit_test

ax3l commented:
It looks like the subs hang on the new test. Maybe you can exclude the dashboard tests from pytest in stubs to overcome this?

proy30 committed:
ignore dashboard test in stubs workflow

ax3l committed:
CMake: Support In-Build Testing

ax3l committed:
GH Actions: In-Build Testing

Ensuring that we continue to support testing
via CTest from the build directory alone, without
the need to first install the Python modules.

ax3l committed:
pytest: Fix Test Skip Logic (Imports)

ax3l reviewed (approved):
This is great, thank you so much!
I finalized the PR and will merge it now (please see the commits I added for details).

Please see the open comment by @EZoni on the wait logic. This can be addressed in a follow-up PR if needed.

ax3l commented:
```
2025-06-22T04:45:02.5764735Z Error copying directory from "/home/runner/work/impactx/impactx/src/python/impactx" to "/home/runner/work/impactx/impactx/build/lib/site-packages/impactx".
```
Uh oh, are we running out of CI disk space? :thinking:

ax3l commented:
@EZoni we might need to port over our WarpX "clean up each test after it ran and succeeded" logic for ctest to ImpactX.

ax3l committed:
CMake: CI w/ `ImpactX_TEST_CLEANUP`

Ported from WarpX

ax3l commented:
Got this, pushed the functionality.

Will enable for debug builds in CI for now, because they create the biggest binary artifacts. But can enable on all runners as well later on.

ax3l commented:
Ah, need to define dependent jobs as in WarpX, too.

ax3l committed:
Fix Solenoid Dependencies 

Cleanup depends on each other

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -681,6 +699,10 @@ add_impactx_test(solenoid.restart
 set_property(TEST solenoid.restart.run APPEND PROPERTY DEPENDS "solenoid.run")
+if(ImpactX_TEST_CLEANUP)
+    # do not clean up dependency test before current test is completed
+    set_property(TEST solenoid.cleanup APPEND PROPERTY DEPENDS "soilenoid.restart.cleanup")
```
```suggestion
    set_property(TEST solenoid.cleanup APPEND PROPERTY DEPENDS "solenoid.restart.cleanup")
```

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -691,6 +713,10 @@ add_impactx_test(solenoid.restart.py
 endif()
+if(ImpactX_TEST_CLEANUP)
+    # do not clean up dependency test before current test is completed
+    set_property(TEST solenoid.py.cleanup APPEND PROPERTY DEPENDS "soilenoid.restart.py.cleanup")
```
```suggestion
    set_property(TEST solenoid.py.cleanup APPEND PROPERTY DEPENDS "solenoid.restart.py.cleanup")
```

ax3l committed:
fix typo last commit

Phone heh

ax3l committed:
Merge remote-tracking branch 'mainline/development' into dashboard/add_first_unit_test

ax3l commented:
Did a few CI updates in separate PRs (#994 #995) -- now this should pass :)

ax3l commented:
Hm, still
```
2025-06-23T07:05:56.5156822Z Run Build Command(s): /usr/local/bin/ninja -v pip_install
...
FAILED: lib/site-packages/impactx/impactx_pybind.cpython-310-x86_64-linux-gnu.so 
...
2025-06-23T07:06:00.2608304Z Error copying directory from "/home/runner/work/impactx/impactx/src/python/impactx" to "/home/runner/work/impactx/impactx/build/lib/site-packages/impactx".
...
2025-06-23T07:06:06.5634664Z ninja: build stopped: subcommand failed.
```

That looks still like running out of space to me...

ax3l committed:
Ubuntu CI: Avoid `sudo` in install

Might change permissions in the build dir...?

ax3l commented:
Ok, that was actually a permission issue from doing
```
sudo ... install
... pip_install
```
Must have changed permissions in `build/` in the first step. fixed.
